packages/protocol/test-sol/integration/Integration.t.sol

```diff
@@ -1,19 +1,77 @@
 // SPDX-License-Identifier: UNLICENSED
-pragma solidity ^0.5.13;
+pragma solidity ^0.8.0;
+
+// import "forge-std/console2.sol";
+
+// import { Constants } from "@test-sol/constants.sol";
+import "celo-foundry-8/Test.sol";
 
-import "celo-foundry/Test.sol";
-import "@celo-contracts/common/GoldToken.sol";
-// import "@celo-contracts/common/test/MockGoldToken.sol";
-import "forge-std/console.sol";
 import "@celo-contracts/common/interfaces/IRegistry.sol";
 
 contract IntegrationTest is Test {
   address constant registryAddress = address(0x000000000000000000000000000000000000ce10);
-  address account1 = actor("account1");
-  address account2 = actor("account2");
   IRegistry registry = IRegistry(registryAddress);
 
-  function setUp() public {}
+  // address account1 = actor("account1");
+  // address account2 = actor("account2");
 
-  function test_dummy() public {}
+  function setUp() public {}
 }
+
+contract RegistryIntegrationTest is IntegrationTest {
+  string[23] expectedContractsInRegistry;
+
+  // TODO(Arthur): Consider moving this to a config file. Perhaps make the migration depend
+  // on that file too?
+  constructor() {
+    expectedContractsInRegistry = [
+      "Accounts",
+      "BlockchainParameters",
+      "DoubleSigningSlasher",
+      "DowntimeSlasher",
+      "Election",
+      "EpochRewards",
+      "Escrow",
+      "FederatedAttestations",
+      "FeeCurrencyWhitelist",
+      "FeeCurrencyDirectory",
+      "Freezer",
+      "FeeHandler",
+      "GoldToken",
+      "Governance",
+      "GovernanceSlasher",
+      "LockedGold",
+      "OdisPayments",
+      "Random",
+      "Registry",
+      "SortedOracles",
+      "UniswapFeeHandlerSeller",
+      "MentoFeeHandlerSeller",
+      "Validators"
+    ];
+  }
+
+  function test_shouldHaveAddressInRegistry() public view {
+    for (uint256 i = 0; i < expectedContractsInRegistry.length; i++) { 
+      string memory contractName = expectedContractsInRegistry[i];
+      address contractAddress = registry.getAddressFor(keccak256(abi.encodePacked(contractName)));
+      console2.log(contractName, "address in Registry is: ", contractAddress);
+      assert(contractAddress != address(0));
+    }
+  }
+
+  function test_shouldHaveCorrectBytecode() public {
+    ////////////////////DEBUGGING: START/////////////////////////
+    // SPECIFIC EXAMPLE REGISTRY.SOL BEFORE LOOPING OVER ALL CONTRACTS
+    string memory contractName = "Registry";
+    address contractAddress = registry.getAddressForStringOrDie(contractName);
+    ////////////////////DEBUGGING: END///////////////////////////
+
+    // Get the deployed bytecode
+    bytes memory actualBytecode = contractAddress.code;
+    bytes memory expectedBytecode = vm.getDeployedCode(string.concat(contractName, ".sol"));
+
+    // Compare the bytecodes
+    assertEq(actualBytecode, expectedBytecode, "Bytecode does not match");
```

Comment:
As of https://github.com/celo-org/celo-monorepo/pull/11002/commits/d99a7465129125ecfd2d9eb66763c6e77cc53623, the actual and expected bytecode doesn't match for `Registry.sol` (as a specific example). 

I'm investigating why that is.

Reply:
As of https://github.com/celo-org/celo-monorepo/pull/11002/commits/4df1e7c514e093852d786d673e1ae976e1f9a92c, I'm getting the runtime bytecode from the implementation address, instead of the proxy address, which is registered in `Registry.sol`. 

Now the bytecode for `Regsitry.sol` (as a specific example) only differ slightly at the end of the bytecode:

```sh
Expected: 0x608060405234801561001057600080fd5b50600436106100cf5760003560e01c80638932cbf41161008c578063c586579311610066578063c58657931461029e578063dcf0aaed14610315578063dd92723314610332578063f2fde38b1461034f576100cf565b80638932cbf4146102205780638da5cb5b1461028e5780638f32d59b14610296576100cf565b8063158ef93e146100d457806317c50818146100f0578063715018a6146101675780637ef50298146101715780638129fc1c146101aa578063853db323146101b2575b600080fd5b6100dc610375565b604080519115158252519081900360200190f35b6100dc6004803603604081101561010657600080fd5b810190602081018135600160201b81111561012057600080fd5b82018360208201111561013257600080fd5b803590602001918460208302840111600160201b8311171561015357600080fd5b9193509150356001600160a01b0316610385565b61016f610403565b005b61018e6004803603602081101561018757600080fd5b50356104a6565b604080516001600160a01b039092168252519081900360200190f35b61016f6104c1565b61018e600480360360208110156101c857600080fd5b810190602081018135600160201b8111156101e257600080fd5b8201836020820111156101f457600080fd5b803590602001918460018302840111600160201b8311171561021557600080fd5b50909250905061053e565b61018e6004803603602081101561023657600080fd5b810190602081018135600160201b81111561025057600080fd5b82018360208201111561026257600080fd5b803590602001918460018302840111600160201b8311171561028357600080fd5b50909250905061058f565b61018e61064e565b6100dc61065d565b61016f600480360360408110156102b457600080fd5b810190602081018135600160201b8111156102ce57600080fd5b8201836020820111156102e057600080fd5b803590602001918460018302840111600160201b8311171561030157600080fd5b9193509150356001600160a01b0316610681565b61018e6004803603602081101561032b57600080fd5b5035610799565b61018e6004803603602081101561034857600080fd5b503561081e565b61016f6004803603602081101561036557600080fd5b50356001600160a01b0316610839565b600054600160a01b900460ff1681565b6000805b838110156103f657826001600160a01b0316600160008787858181106103ab57fe5b60209081029290920135835250810191909152604001600020546001600160a01b031614156103de5760019150506103fc565b6103ef81600163ffffffff61089e16565b9050610389565b50600090505b9392505050565b61040b61065d565b61045c576040805162461bcd60e51b815260206004820181905260248201527f4f776e61626c653a2063616c6c6572206973206e6f7420746865206f776e6572604482015290519081900360640190fd5b600080546040516001600160a01b03909116907f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e0908390a3600080546001600160a01b0319169055565b6001602052600090815260409020546001600160a01b031681565b600054600160a01b900460ff1615610520576040805162461bcd60e51b815260206004820152601c60248201527f636f6e747261637420616c726561647920696e697469616c697a656400000000604482015290519081900360640190fd5b6000805460ff60a01b1916600160a01b17905561053c336108f8565b565b6000808383604051602001808383808284376040805191909301818103601f19018252835280516020918201206000908152600190915291909120546001600160a01b031698975050505050505050565b6000808383604051602001808383808284376040805191909301818103601f19018252835280516020918201206000818152600190925292902054919550506001600160a01b03161515925061062f915050576040805162461bcd60e51b815260206004820181905260248201527f6964656e74696669657220686173206e6f20726567697374727920656e747279604482015290519081900360640190fd5b6000908152600160205260409020546001600160a01b03169392505050565b6000546001600160a01b031690565b600080546001600160a01b0316610672610998565b6001600160a01b031614905090565b61068961065d565b6106da576040805162461bcd60e51b815260206004820181905260248201527f4f776e61626c653a2063616c6c6572206973206e6f7420746865206f776e6572604482015290519081900360640190fd5b60008383604051602001808383808284376040805191909301818103601f19018252808452815160209283012060008181526001845285902080546001600160a01b0319166001600160a01b038c169081179091558383529282018b905297509095508694507f4166d073a7a5e704ce0db7113320f88da2457f872d46dc020c805c562c1582a09350899289925081908101848480828437600083820152604051601f909101601f19169092018290039550909350505050a350505050565b6000818152600160205260408120546001600160a01b0316610802576040805162461bcd60e51b815260206004820181905260248201527f6964656e74696669657220686173206e6f20726567697374727920656e747279604482015290519081900360640190fd5b506000908152600160205260409020546001600160a01b031690565b6000908152600160205260409020546001600160a01b031690565b61084161065d565b610892576040805162461bcd60e51b815260206004820181905260248201527f4f776e61626c653a2063616c6c6572206973206e6f7420746865206f776e6572604482015290519081900360640190fd5b61089b816108f8565b50565b6000828201838110156103fc576040805162461bcd60e51b815260206004820152601b60248201527f536166654d6174683a206164646974696f6e206f766572666c6f770000000000604482015290519081900360640190fd5b6001600160a01b03811661093d5760405162461bcd60e51b815260040180806020018281038252602681526020018061099d6026913960400191505060405180910390fd5b600080546040516001600160a01b03808516939216917f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e091a3600080546001600160a01b0319166001600160a01b0392909216919091179055565b339056fe4f776e61626c653a206e6577206f776e657220697320746865207a65726f2061646472657373a265627a7a72315820165a229e90bd9ca2de3bbc2b619fe25077a0f031068a0dee44f2fc2f2cfa1eb264736f6c63430005110032

Actual: 0x608060405234801561001057600080fd5b50600436106100cf5760003560e01c80638932cbf41161008c578063c586579311610066578063c58657931461029e578063dcf0aaed14610315578063dd92723314610332578063f2fde38b1461034f576100cf565b80638932cbf4146102205780638da5cb5b1461028e5780638f32d59b14610296576100cf565b8063158ef93e146100d457806317c50818146100f0578063715018a6146101675780637ef50298146101715780638129fc1c146101aa578063853db323146101b2575b600080fd5b6100dc610375565b604080519115158252519081900360200190f35b6100dc6004803603604081101561010657600080fd5b810190602081018135600160201b81111561012057600080fd5b82018360208201111561013257600080fd5b803590602001918460208302840111600160201b8311171561015357600080fd5b9193509150356001600160a01b0316610385565b61016f610403565b005b61018e6004803603602081101561018757600080fd5b50356104a6565b604080516001600160a01b039092168252519081900360200190f35b61016f6104c1565b61018e600480360360208110156101c857600080fd5b810190602081018135600160201b8111156101e257600080fd5b8201836020820111156101f457600080fd5b803590602001918460018302840111600160201b8311171561021557600080fd5b50909250905061053e565b61018e6004803603602081101561023657600080fd5b810190602081018135600160201b81111561025057600080fd5b82018360208201111561026257600080fd5b803590602001918460018302840111600160201b8311171561028357600080fd5b50909250905061058f565b61018e61064e565b6100dc61065d565b61016f600480360360408110156102b457600080fd5b810190602081018135600160201b8111156102ce57600080fd5b8201836020820111156102e057600080fd5b803590602001918460018302840111600160201b8311171561030157600080fd5b9193509150356001600160a01b0316610681565b61018e6004803603602081101561032b57600080fd5b5035610799565b61018e6004803603602081101561034857600080fd5b503561081e565b61016f6004803603602081101561036557600080fd5b50356001600160a01b0316610839565b600054600160a01b900460ff1681565b6000805b838110156103f657826001600160a01b0316600160008787858181106103ab57fe5b60209081029290920135835250810191909152604001600020546001600160a01b031614156103de5760019150506103fc565b6103ef81600163ffffffff61089e16565b9050610389565b50600090505b9392505050565b61040b61065d565b61045c576040805162461bcd60e51b815260206004820181905260248201527f4f776e61626c653a2063616c6c6572206973206e6f7420746865206f776e6572604482015290519081900360640190fd5b600080546040516001600160a01b03909116907f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e0908390a3600080546001600160a01b0319169055565b6001602052600090815260409020546001600160a01b031681565b600054600160a01b900460ff1615610520576040805162461bcd60e51b815260206004820152601c60248201527f636f6e747261637420616c726561647920696e697469616c697a656400000000604482015290519081900360640190fd5b6000805460ff60a01b1916600160a01b17905561053c336108f8565b565b6000808383604051602001808383808284376040805191909301818103601f19018252835280516020918201206000908152600190915291909120546001600160a01b031698975050505050505050565b6000808383604051602001808383808284376040805191909301818103601f19018252835280516020918201206000818152600190925292902054919550506001600160a01b03161515925061062f915050576040805162461bcd60e51b815260206004820181905260248201527f6964656e74696669657220686173206e6f20726567697374727920656e747279604482015290519081900360640190fd5b6000908152600160205260409020546001600160a01b03169392505050565b6000546001600160a01b031690565b600080546001600160a01b0316610672610998565b6001600160a01b031614905090565b61068961065d565b6106da576040805162461bcd60e51b815260206004820181905260248201527f4f776e61626c653a2063616c6c6572206973206e6f7420746865206f776e6572604482015290519081900360640190fd5b60008383604051602001808383808284376040805191909301818103601f19018252808452815160209283012060008181526001845285902080546001600160a01b0319166001600160a01b038c169081179091558383529282018b905297509095508694507f4166d073a7a5e704ce0db7113320f88da2457f872d46dc020c805c562c1582a09350899289925081908101848480828437600083820152604051601f909101601f19169092018290039550909350505050a350505050565b6000818152600160205260408120546001600160a01b0316610802576040805162461bcd60e51b815260206004820181905260248201527f6964656e74696669657220686173206e6f20726567697374727920656e747279604482015290519081900360640190fd5b506000908152600160205260409020546001600160a01b031690565b6000908152600160205260409020546001600160a01b031690565b61084161065d565b610892576040805162461bcd60e51b815260206004820181905260248201527f4f776e61626c653a2063616c6c6572206973206e6f7420746865206f776e6572604482015290519081900360640190fd5b61089b816108f8565b50565b6000828201838110156103fc576040805162461bcd60e51b815260206004820152601b60248201527f536166654d6174683a206164646974696f6e206f766572666c6f770000000000604482015290519081900360640190fd5b6001600160a01b03811661093d5760405162461bcd60e51b815260040180806020018281038252602681526020018061099d6026913960400191505060405180910390fd5b600080546040516001600160a01b03808516939216917f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e091a3600080546001600160a01b0319166001600160a01b0392909216919091179055565b339056fe4f776e61626c653a206e6577206f776e657220697320746865207a65726f2061646472657373a265627a7a72315820613048cebb9685168ef640c1d8aeddf9ec489b7d56164f766881aed6c593e69c64736f6c63430005110032
```

Reply:
As of , I'm still getting this error

I tried with the `Freezer` contract as a specific example:

```sh
    Expected: 0x608060405234801561001057600080fd5b50600436106100935760003560e01c80638d1fdf2f116100665780638d1fdf2f146100ec5780638da5cb5b146101125780638f32d59b14610136578063e58398361461013e578063f2fde38b1461016457610093565b8063158ef93e1461009857806345c8b1a6146100b4578063715018a6146100dc5780638129fc1c146100e4575b600080fd5b6100a061018a565b604080519115158252519081900360200190f35b6100da600480360360208110156100ca57600080fd5b50356001600160a01b031661019a565b005b6100da610202565b6100da610293565b6100da6004803603602081101561010257600080fd5b50356001600160a01b0316610310565b61011a61037e565b604080516001600160a01b039092168252519081900360200190f35b6100a061038d565b6100a06004803603602081101561015457600080fd5b50356001600160a01b03166103b1565b6100da6004803603602081101561017a57600080fd5b50356001600160a01b03166103c6565b600054600160a01b900460ff1681565b6101a261038d565b6101e1576040805162461bcd60e51b815260206004820181905260248201526000805160206104e4833981519152604482015290519081900360640190fd5b6001600160a01b03166000908152600160205260409020805460ff19169055565b61020a61038d565b610249576040805162461bcd60e51b815260206004820181905260248201526000805160206104e4833981519152604482015290519081900360640190fd5b600080546040516001600160a01b03909116907f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e0908390a3600080546001600160a01b0319169055565b600054600160a01b900460ff16156102f2576040805162461bcd60e51b815260206004820152601c60248201527f636f6e747261637420616c726561647920696e697469616c697a656400000000604482015290519081900360640190fd5b6000805460ff60a01b1916600160a01b17905561030e33610419565b565b61031861038d565b610357576040805162461bcd60e51b815260206004820181905260248201526000805160206104e4833981519152604482015290519081900360640190fd5b6001600160a01b03166000908152600160208190526040909120805460ff19169091179055565b6000546001600160a01b031690565b600080546001600160a01b03166103a26104b9565b6001600160a01b031614905090565b60016020526000908152604090205460ff1681565b6103ce61038d565b61040d576040805162461bcd60e51b815260206004820181905260248201526000805160206104e4833981519152604482015290519081900360640190fd5b61041681610419565b50565b6001600160a01b03811661045e5760405162461bcd60e51b81526004018080602001828103825260268152602001806104be6026913960400191505060405180910390fd5b600080546040516001600160a01b03808516939216917f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e091a3600080546001600160a01b0319166001600160a01b0392909216919091179055565b339056fe4f776e61626c653a206e6577206f776e657220697320746865207a65726f20616464726573734f776e61626c653a2063616c6c6572206973206e6f7420746865206f776e6572a265627a7a723158200c7ee4bb68f4c59bf66e931fb4a8100a08937c95dc474145838c9f6ed56e9b7664736f6c63430005110032
      Actual: 0x608060405234801561001057600080fd5b50600436106100935760003560e01c80638d1fdf2f116100665780638d1fdf2f146100ec5780638da5cb5b146101125780638f32d59b14610136578063e58398361461013e578063f2fde38b1461016457610093565b8063158ef93e1461009857806345c8b1a6146100b4578063715018a6146100dc5780638129fc1c146100e4575b600080fd5b6100a061018a565b604080519115158252519081900360200190f35b6100da600480360360208110156100ca57600080fd5b50356001600160a01b031661019a565b005b6100da610202565b6100da610293565b6100da6004803603602081101561010257600080fd5b50356001600160a01b0316610310565b61011a61037e565b604080516001600160a01b039092168252519081900360200190f35b6100a061038d565b6100a06004803603602081101561015457600080fd5b50356001600160a01b03166103b1565b6100da6004803603602081101561017a57600080fd5b50356001600160a01b03166103c6565b600054600160a01b900460ff1681565b6101a261038d565b6101e1576040805162461bcd60e51b815260206004820181905260248201526000805160206104e4833981519152604482015290519081900360640190fd5b6001600160a01b03166000908152600160205260409020805460ff19169055565b61020a61038d565b610249576040805162461bcd60e51b815260206004820181905260248201526000805160206104e4833981519152604482015290519081900360640190fd5b600080546040516001600160a01b03909116907f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e0908390a3600080546001600160a01b0319169055565b600054600160a01b900460ff16156102f2576040805162461bcd60e51b815260206004820152601c60248201527f636f6e747261637420616c726561647920696e697469616c697a656400000000604482015290519081900360640190fd5b6000805460ff60a01b1916600160a01b17905561030e33610419565b565b61031861038d565b610357576040805162461bcd60e51b815260206004820181905260248201526000805160206104e4833981519152604482015290519081900360640190fd5b6001600160a01b03166000908152600160208190526040909120805460ff19169091179055565b6000546001600160a01b031690565b600080546001600160a01b03166103a26104b9565b6001600160a01b031614905090565b60016020526000908152604090205460ff1681565b6103ce61038d565b61040d576040805162461bcd60e51b815260206004820181905260248201526000805160206104e4833981519152604482015290519081900360640190fd5b61041681610419565b50565b6001600160a01b03811661045e5760405162461bcd60e51b81526004018080602001828103825260268152602001806104be6026913960400191505060405180910390fd5b600080546040516001600160a01b03808516939216917f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e091a3600080546001600160a01b0319166001600160a01b0392909216919091179055565b339056fe4f776e61626c653a206e6577206f776e657220697320746865207a65726f20616464726573734f776e61626c653a2063616c6c6572206973206e6f7420746865206f776e6572a265627a7a7231582081345f464357bec1a1fcfa537e4f65331dc46a48129550091a043b887d26815164736f6c63430005110032
```

Reply:
Unless there is a better problem to work on, I'll try to
1. come up with a way to compare the actual and expected bytecode (this might help understand what's different)
1. parse the artifact bytecode from the `.json` file in the `/build` directory instead of using the Foundry helper which is using the `/out` directory.

Reply:
For some contracts like `Validators.sol`, `getDeployedCode("Validators.sol")` doesn't work with the following error:

```sh
    ├─ [0] VM::getDeployedCode("Validators.sol") [staticcall]
    │   └─ ← data did not match any variant of untagged enum Bytecode at line 2312 column 3
    └─ ← data did not match any variant of untagged enum Bytecode at line 2312 column 3
```

I'm not sure what the reason for the bug is.

Reply:
### Explanation:

The bytecodes differ because the last bytes include a Swarm or IPFS hash:

> The compiler currently by default appends the [IPFS hash (in CID v0)](https://docs.ipfs.tech/concepts/content-addressing/#version-0-v0) of the canonical metadata file and the compiler version to the end of the bytecode. Optionally, a Swarm hash instead of the IPFS, or an experimental flag is used. Below are all the possible fields:
> 
> ```json
> {
>   "ipfs": "<metadata hash>",
>   // If "bytecodeHash" was "bzzr1" in compiler settings not "ipfs" but "bzzr1"
>   "bzzr1": "<metadata hash>",
>   // Previous versions were using "bzzr0" instead of "bzzr1"
>   "bzzr0": "<metadata hash>",
>   // If any experimental features that affect code generation are used
>   "experimental": true,
>   "solc": "<compiler version>"
> }
> ```
> 
> Because we might support other ways to retrieve the metadata file in the future, this information is stored [CBOR](https://tools.ietf.org/html/rfc7049)-encoded. The last two bytes in the bytecode indicate the length of the CBOR encoded information. By looking at this length, the relevant part of the bytecode can be decoded with a CBOR decoder.
> 
> Whereas release builds of solc use a 3 byte encoding of the version as shown above (one byte each for major, minor and patch version number), pre-release builds will instead use a complete version string including commit hash and build date.
> 
> The commandline flag --no-cbor-metadata can be used to skip metadata from getting appended at the end of the deployed bytecode. Equivalently, the boolean field settings.metadata.appendCBOR in Standard JSON input can be set to false.

Source: Solidity docs > [Encoding of the Metadata Hash in the Bytecode](https://docs.soliditylang.org/en/develop/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode)

### Foundry fix

> disable metadata with `cbor_metadata = false.`

Source: https://github.com/foundry-rs/foundry/issues/7826

Reply:
### Custom fix

I implemented a helper function in https://github.com/celo-org/celo-monorepo/pull/11002/commits/1e61942c5cf03a1addb8a1fe279fe61c567fa20d that removes the CBOR encoded metadata from a given bytecode before it is compared.

With this change, most contracts pass.

Reply:
As of https://github.com/celo-org/celo-monorepo/pull/11002/commits/e55c973bdd9d3330fbcd06ff58f0765fd3981b76, most contracts pass the bytecode comparison test, apart from:
1. Account 
2. Election
3. Escrow
4. FederatedAttestations
5. Governance
6. SortedOracles
7. Validators

They fail with an error of the form: 

```
    ├─ [0] VM::getDeployedCode("Election.sol") [staticcall]
    │   └─ ← data did not match any variant of untagged enum Bytecode at line 1883 column 3
    └─ ← data did not match any variant of untagged enum Bytecode at line 1883 column 3
```

```
    ├─ [0] VM::getDeployedCode("FederatedAttestations.sol") [staticcall]
    │   └─ ← data did not match any variant of untagged enum Bytecode at line 721 column 3
    └─ ← data did not match any variant of untagged enum Bytecode at line 721 column 3
```

```
    ├─ [0] VM::getDeployedCode("Governance.sol") [staticcall]
    │   └─ ← data did not match any variant of untagged enum Bytecode at line 2705 column 3
    └─ ← data did not match any variant of untagged enum Bytecode at line 2705 column 3
```

```
    ├─ [0] VM::getDeployedCode("SortedOracles.sol") [staticcall]
    │   └─ ← data did not match any variant of untagged enum Bytecode at line 1013 column 3
    └─ ← data did not match any variant of untagged enum Bytecode at line 1013 column 3
```

```
    ├─ [0] VM::getDeployedCode("Validators.sol") [staticcall]
    │   └─ ← data did not match any variant of untagged enum Bytecode at line 2312 column 3
    └─ ← data did not match any variant of untagged enum Bytecode at line 2312 column 3
```

Reply:
My current thinking is that these contracts fail because they depend on linked libraries.

We know the linked libraries are:

```
contracts/common/linkedlists/AddressSortedLinkedListWithMedian.sol
contracts/common/Signatures.sol:Signatures
contracts/common/linkedlists/AddressLinkedList.sol
contracts/common/linkedlists/AddressSortedLinkedList.sol
contracts/common/linkedlists/IntegerSortedLinkedList.sol
contracts/governance/Proposals.sol
```

We know:

1. `Accounts` depends on:

    ```
    import "../common/Signatures.sol";
    ```

1. `Election` depends on:

    ```
    import "../common/linkedlists/AddressSortedLinkedList.sol";
    ```

1. `Escrow` depends on:

    ```
    import "../common/Signatures.sol";
    ```

1. `FederatedAttestations` depends on:
    
    ```
    import "../common/Signatures.sol";
    ```

1. `Governance` depends on:

    ```
    import "../common/linkedlists/IntegerSortedLinkedList.sol";
    ```

1. `SortedOracles` depends on:

    ```
    import "../common/linkedlists/AddressSortedLinkedListWithMedian.sol";
    ```

1. `Validators` depends on:

    ```
    import "../common/linkedlists/AddressLinkedList.sol";
    ```

This is a known limitation of Foundry at the moment: 
- https://github.com/foundry-rs/foundry/issues/6120

Reply:
I'm comfortable excluding these contracts from the bytecode test until this is better supported in Foundry.
This is done in https://github.com/celo-org/celo-monorepo/pull/11002/commits/e55c973bdd9d3330fbcd06ff58f0765fd3981b76.